### PR TITLE
chore(docs): add requirements re ports and stun server to docs

### LINKED ISSUE #11876: chore(docs): update /networking with information on port requirements
Our docs do not make any mention of the port requirements for Coder, either for direct or indirect connections. 
Provide:
- A high-level summary of all ports required for direct connections.
- A high-level summary of how a direct connection is established to a Coder workspace.

An in-depth explanation of STUN / DERP is out of scope here. Just link to relevant documentation if folks want to read more.

### PR DESCRIPTION
**Pull request by johnstcn**
Fixes https://github.com/coder/coder/issues/11876

### CODE REVIEW

File: docs/networking/index.md [hunk 1]
```diff
@@ -13,6 +13,37 @@ user <-> workspace connections are end-to-end encrypted.
 
 [Tailscale's open source](https://tailscale.com) backs our networking logic.
 
+## Requirements
+
+In order for clients and workspaces to be able to connect:
+
+- All clients and agents must be able to establish a connection to the Coder
+  server (`CODER_ACCESS_URL`) over HTTPS (`tcp/443`).
```
Comment (deansheather) [inline docs/networking/index.md:21]:
```suggestion
  server (`CODER_ACCESS_URL`) over HTTP/HTTPS.
```

Isn't necessarily tcp/443, custom ports are allowed.

File: docs/networking/index.md [hunk 2]
```diff
@@ -13,6 +13,37 @@ user <-> workspace connections are end-to-end encrypted.
...
+- The client is connecting using the CLI (e.g. `coder ssh` or
+  `coder port-forward`).
+- The client and workspace agent are both able to connect to a specific STUN
+  server.
+  > The STUN server needs to tell the client and workspace their respective
+  > `address:port` pairs from its perspective so that they can establish a
+  > direct connection with each other. If the client and agent are only able to
+  > connect to STUN servers on different networks, then a direct connection will
+  > not be possible. For an in-depth technical explanation, see
+  > [How NAT traversal works (tailscale.com)](https://tailscale.com/blog/how-nat-traversal-works).
```
Comment (deansheather) [inline docs/networking/index.md:41]: Technically, both need to be able to send packets and receive packets from a STUN server in the same network "zone". No connections since it's UDP

File: docs/networking/index.md [hunk 3]
```diff
@@ -13,6 +13,37 @@ user <-> workspace connections are end-to-end encrypted.
...
+- Outbound UDP traffic must be allowed for both the client and the agent from
+  source ports `udp/3478` and `udp/41641` to all destination ports.
+  > For more detailed information, see
+  > [What firewall ports should I open to use Tailscale? (tailscale.com)](https://tailscale.com/kb/1082/firewall-ports).
```
Comment (deansheather) [inline docs/networking/index.md:45]: This is only true for tailscale and not for Coder AFAIK. We listen on a random port for our UDP socket

File: docs/networking/index.md [hunk 4]
```diff
@@ -13,6 +13,37 @@ user <-> workspace connections are end-to-end encrypted.
...
+
```
Comment (deansheather) [inline docs/networking/index.md:46]: There should probably be some details about NATs and stuff but I really don't know what to write without getting super technical :/

File: docs/networking/index.md (hunk 3 above)
Comment (johnstcn) [reply docs/networking/index.md:45]: So essentially all outbound UDP traffic needs to be allowed unconditionally then?

File: docs/networking/index.md (hunk 4 above)
Comment (johnstcn) [reply docs/networking/index.md:46]: Honestly I'd prefer to just link to Tailscale's docs on this.

File: docs/networking/index.md (hunk 3 above)
Comment (deansheather) [reply docs/networking/index.md:45]: Yes, to whatever CIDRs the peers should be allowed to use. And it's both ways, so both firewalls need to allow the other peer

Comment (deansheather) [reply docs/networking/index.md:45]: The reason we use a random port is that multiple CLI connections can be open simultaneously and they don't share a daemon (like in Tailscale's situation)

File: docs/networking/index.md (hunk 4 above)
Comment (deansheather) [reply docs/networking/index.md:46]: IDK how well Tailscale's docs cut it, this one doesn't really explain much about NAT just about firewalls.

Comment (johnstcn) [reply docs/networking/index.md:46]: I link to https://tailscale.com/blog/how-nat-traversal-works a bit above in the STUN section.

Comment (deansheather) [reply docs/networking/index.md:46]: This is a good blog post but it's not very good documentation since it's 9000 words long. We should probably dumb it down

Comment (johnstcn) [reply docs/networking/index.md:61]: OK, I added a fairly high-level overview.

Comment (spikecurtis) [reply docs/networking/index.md:92]: I've Slack'd you some diagrams.  If we're going to explain it, then those are my suggested examples.

File: docs/networking/stun.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+  vpn1 === vpn2
+  vpn2 --> stun2
+  client === homenat
+  homenat === vpn1
+  homenat x-.-x stun1
+  agent --- clusternat
+  clusternat --- corpnat
+  corpnat --> stun1
+  corpnat --> stun2
+```
```
Comment (johnstcn) [inline docs/networking/stun.md:193]:
```mermaid
flowchart TD
  subgraph homenet["Home Network"]
    client["Client workstation\n192.168.1.101"]
    homenat["Home Router/NAT\n65.4.3.21"]
  end
  subgraph internet["Public Internet"]
    stun1["Public STUN"]
    vpn1["VPN entry node"]
  end
  subgraph corpnet["Corp Network 172.16.1.0/24"]
    corpnat["Corp Router/NAT\n172.16.1.1\n12.34.56.7"]
    vpn2["VPN exit node\n172.16.1.2"]
    stun2["Private STUN"]
    subgraph cluster["Cluster Network 10.11.12.0/16"]
      clusternat["Cluster Router/NAT\n10.11.12.1\n172.16.1.254"]
      agent["Workspace agent\n10.11.12.34"]
    end
  end
  vpn1 === vpn2
  vpn2 --> stun2
  client === homenat
  homenat === vpn1
  homenat x-.-x stun1
  agent --- clusternat
  clusternat --- corpnat
  corpnat --> stun1
  corpnat --> stun2
```

File: docs/networking/stun.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+  end
+
+  subgraph internet["Public Internet"]
+  end
+
+  client -- "[12.34.56.7:41563]" --- homenat
+  homenat -- "[12.34.56.7:41563]" --- internet
+  internet -- "[12.34.56.7:41563]" --- corpnat
+  corpnat -- "[10.21.43.241:56812]" --> agent
+```
```
Comment (johnstcn) [inline docs/networking/stun.md:128]:
```mermaid
flowchart LR
  subgraph homenet["Home Network"]
    direction LR
    client["Client workstation\n192.168.1.101:38297"]
    homenat["Home Router/NAT\n65.4.3.21:29187"]
  end
  subgraph corpnet["Corp Network"]
    direction LR
    agent["Workspace agent\n10.21.43.241:56812"]
    corpnat["Corp Router/NAT\n12.34.56.7:41563"]
  end
  subgraph internet["Public Internet"]
  end
  client -- "[12.34.56.7:41563]" --- homenat
  homenat -- "[12.34.56.7:41563]" --- internet
  internet -- "[12.34.56.7:41563]" --- corpnat
  corpnat -- "[10.21.43.241:56812]" --> agent
```

File: docs/networking/stun.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+  subgraph corpnet["Network B"]
+    agent["Workspace agent\n10.21.43.241:56812"]
+    corpnat["NAT\n??.??.??.??:?????"]
+  end
+
+  client --- homenat
+  agent --- corpnat
+  corpnat -- "[I see 12.34.56.7:41563]" --> stun1
+  homenat -- "[I see 65.4.3.21:29187]" --> stun1
+```
```
Comment (johnstcn) [inline docs/networking/stun.md:103]:
```mermaid
flowchart LR
  subgraph homenet["Network A"]
    client["Client workstation\n192.168.1.101:38297"]
    homenat["NAT\n??.??.??.??:?????"]
  end
  subgraph internet["Public Internet"]
    stun1["STUN server"]
  end
  subgraph corpnet["Network B"]
    agent["Workspace agent\n10.21.43.241:56812"]
    corpnat["NAT\n??.??.??.??:?????"]
  end
  client --- homenat
  agent --- corpnat
  corpnat -- "[I see 12.34.56.7:41563]" --> stun1
  homenat -- "[I see 65.4.3.21:29187]" --> stun1
```

File: docs/networking/stun.md [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+### 1. Direct connections without NAT or STUN
+
+```mermaid
+flowchart LR
+    subgraph corpnet["Private Network\ne.g. Corp. LAN"]
+    A[Client Workstation\n192.168.21.47:38297]
+    C[Workspace Agent\n192.168.21.147:41563]
+    A <--> C
+    end
+```
```
Comment (johnstcn) [inline docs/networking/stun.md:76]:
```mermaid
flowchart LR
    subgraph corpnet["Private Network\ne.g. Corp. LAN"]
    A[Client Workstation\n192.168.21.47:38297]
    C[Workspace Agent\n192.168.21.147:41563]
    A <--> C
    end
```

File: docs/networking/index.md (hunk 4 above)
Comment (johnstcn) [reply docs/networking/index.md:58]: Added and moved to a separate page as this one is getting plenty big already.

**Review by spikecurtis — APPROVED** [review]
This is great!  Couple small language tweaks and suggestions inline

File: docs/networking/stun.md (hunk 6 above)
Comment (johnstcn) [reply docs/networking/stun.md:127]: For now, fixed by changing the arrow directions.
